_posts/on-working-with-viewmodel-composition-based-system.md

```diff
@@ -0,0 +1,145 @@
+
+That begin said, it's not gold all that glitters. Setting up integration testing is way more complicated when using different repositories for each component. It's not impossible. However, we'll have to spend time designing custom components and tools specifically aimed at supporting integration tests.
+
+Last but not least. Even if it's true that using a repository for each component is a cleaner approach, it'll introduce friction into the development process, especially at the early stages. Let's imagine that we're adding a new restful API. The first step is to change the restful API repository; then, we need to publish the changes so that downstream components can consume them. The development effort then moves onto composition handlers. Update the dependency on the restful API package to have integration tests fail or be able to create the new required tests. Adjust the composition handlers to support the new restful API and publish the package. Apply, if needed, a similar process to the frontend component.
+
+> For more details about integration testing when using HTTP endpoints, take a look at ["You wanna test HTTP, right?"](https://milestone.topics.it/2021/04/28/you-wanna-test-http.html) and the latest [HTTP client and HTTP client factory in integration testing](link).
+
+I wouldn't define the described process as smooth.
+
+As for the previous approach, let's take a look at the pros and cons:
+
+- Cons: the overall development process will come with more frictions
+- Cons: multiple components integration testing will be more challenging
+- Pros: simpler CI/CD pipelines
+- Pros: each release will release only the changed components
+
+Sounds like exactly the opposite of a repository for each logical service. 
+
+From the structure perspective, a few other approaches might sit in the middle of what we described so far. I guess that they would also lead to a similar distribution of pros and cons.
+
+## Shall we blindly use ViewModel Composition?
+
+The quick answer is obviously no. And not only for the technical challenges presented in this article.
+
+There are many factors that we should consider before deciding to use ViewModel Composition. For example, the company size, team competencies, and the system's overall complexity are all good metrics to look at before making any final decision. ViewModel Composition will only make things more complex if the company or project is small and the team is not so experienced. It'd be better to design frontends using monolithic architecture. One frontend repository contains all the code and artifacts and knows how to communicate with the distributed systems backend.
+
+It's probably fair to say that a composite UI is the last of the problems; not necessarily the least important, but the last to address.
+
+We tend to think in terms of consequences. The reasoning usually goes like this: if we design that component in a suboptimal way when we'll fix it, it will hurt all these other components.
```

Comment:
```suggestion
We tend to think in terms of consequences. The reasoning usually goes like this: if we design that component in a suboptimal way, then when we fix it, it will hurt all these other components.
```